Switch to tri-color marking GC

ke-yu opened the pull request:
### Purpose

Current DesignScript virtual machine uses [naive mark and sweep GC](https://en.wikipedia.org/wiki/Tracing_garbage_collection#Na.C3.AFve_mark-and-sweep). The major problem of naive mark and sweep is the system will be suspended during collection to examine the whole memory. Because of this disadvantage, we can't trigger GC too frequently, otherwise the system might be frozen for a long time if there are some big objects. Instead, GC is only is triggered after each `POP` operation at top-level associative language block. And as we try to keep the implementation as simple as possible, GC even won't be triggered in function call, therefore memory consumption could easily go up for some high-order function node like `List.Map`. 

This PR switches naive mark and sweep to [tri-color marking] (https://en.wikipedia.org/wiki/Tracing_garbage_collection#Tri-color_marking). This method won't stop the whole world when doing GC. Instead, it could be done step by step, and in each step it could only mark one or few objects.

This PR also enhance `Heap` so that GC is more configurable. For example, GC will be triggered based on memory allocation condition. 

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files

### Performance Benchmark

I run ProtoTest, DynamoCoreTests and DynamoCoreWpfTests, and do memory sampling for every 100ms with different GC methods. Following graph shows performance benchmark result:

![mem1](https://cloud.githubusercontent.com/assets/2600379/8493062/218b6d2e-2187-11e5-893c-482bb7047985.png)

![mem2](https://cloud.githubusercontent.com/assets/2600379/8493063/24133c3e-2187-11e5-84c9-c87b5487f9d1.png)

![mem4](https://cloud.githubusercontent.com/assets/2600379/8518616/5b6f8dda-23fe-11e5-943c-cf8f76107dd5.png)

Both GC methods' memory footprint are close, though tri-color marking has little higher memory usage for ProtoTest. 

For these three tests, tri-color marking takes slightly shorter execution time, which primarily is because the heap now monitors its memory allocation and only triggers GC when memory usage is over some threshold. 

For frequent memory allocation, for example `DSCore_LaceLong` test case:

![longest-lacing](https://cloud.githubusercontent.com/assets/2600379/8493672/112b9676-2190-11e5-8bf4-d8ff326e61bb.png)

The inputs are 1000 and 10000. Following graph shows the result:

![memory3](https://cloud.githubusercontent.com/assets/2600379/8493673/1d32d6dc-2190-11e5-8131-6d65e2e872b9.png)

Naive mark & sweep consumes over 2GB memory, but tri-color GC keeps very low memory footprint. Although tri-color GC would potentially take longer time, for example, tri-color (64K) takes over 2x longer, but as heap now is configurable, one we increase the GC threshold (when memory is over this limit, GC will be triggered), we get very close execution time and at the same time low memory footprint.

### Reviewers

@lukechurch, @junmendoza  PTAL. Regressions have been fixed.

lukechurch commented on src/Engine/ProtoCore/DSASM/Executive.cs:
```diff
@@ -4563,9 +4563,12 @@ protected StackValue POP_helper(Instruction instruction, out int blockId, out in
+                PopToIndexedArray(blockId, (int)instruction.op1.opdata, (int)instruction.op2.opdata, dimList, svData);
             }
 
+#if TRACING_GC
```
Is there a reason to run this off by default? Can we invert this so that everyone is getting the new GC?

lukechurch commented on src/Engine/ProtoCore/DSASM/Executive.cs:
```diff
@@ -6849,24 +6867,26 @@ private void Exec(Instruction instruction)
                 rmem.CurrentStackFrame.StackFrameType != StackFrameType.kTypeLanguage)
             {
-                return;
+                return null;
```
What does null signal rather than an empty list?

lukechurch commented on src/Engine/ProtoCore/DSASM/Executive.cs:
```diff
@@ -6849,24 +6867,26 @@ private void Exec(Instruction instruction)
-            var gcRoots = new List<StackValue>();
 
             // Now garbage collection only happens on the top most block. 
             // We will loose this limiation soon.
```
We must have a tracking issue for this limtiation, even if it's old

lukechurch commented on src/Engine/ProtoCore/DSASM/Executive.cs:
```diff
@@ -6849,24 +6867,26 @@ private void Exec(Instruction instruction)
         }
 
-        private void GC()
+        public List<StackValue> CollectRootPointers()
```
Public seems rather broad for this. Is there a reason why this should be part of our API?

lukechurch commented on src/Engine/ProtoCore/DSASM/Executive.cs:
```diff
@@ -6962,14 +6982,21 @@ private void GC()
                 blockId = stackFrame.FunctionCallerBlock;
                 currentFramePointer = stackFrame.FramePointer;
             }
+#endif
```
Is this legal, two consequentives #endifs ? Is this a merge error?

lukechurch commented on src/Engine/ProtoCore/DSASM/Executive.cs:
```diff
@@ -6962,14 +6982,21 @@ private void GC()
 
-            gcRoots.Add(RX);
+#if TRACING_GC
+            gcRoots.AddRange(runtimeCore.ReplicationRoots.Where(r => r.IsReferenceType));
```
This is a fairly inefficient style of coding that LINQ encourages. Is this a hot path? Should we look at storing them in a separate list?

lukechurch commented on src/Engine/ProtoCore/DSASM/Heap.cs:
```diff
@@ -1,15 +1,27 @@
-
 namespace ProtoCore.DSASM
 {
+    public enum GCState
```
public?

lukechurch commented on src/Engine/ProtoCore/DSASM/Heap.cs:
```diff
@@ -245,11 +272,26 @@ public class Heap
-        private bool isGarbageCollecting = false;
+        private readonly StringTable stringTable = new StringTable();
+
+        private const int kGCThreshold = 65536;
```
Lets use .NET STANDARD_FOR_CONSTS.

lukechurch commented on src/Engine/ProtoCore/DSASM/Heap.cs:
```diff
@@ -322,7 +364,7 @@ public StackValue AllocatePointer(int size, MetaData metadata)
         /// Allocate an object pointer.
         /// </summary>
-        /// <param name="size"></param>
+        /// <param name="size"></parame
```
Merge conflict?

lukechurch commented on src/Engine/ProtoCore/DSASM/Heap.cs:
```diff
@@ -520,14 +570,262 @@ private void GCDisposeObject(StackValue svPtr, Executive exe)
+        /// Mark all items in the array.
+        /// </summary>
+        /// <param name="hp"></param>
+        /// <returns></returns>
```
What does it return?

lukechurch commented on src/Engine/ProtoCore/DSASM/Heap.cs:
```diff
@@ -520,14 +570,262 @@ private void GCDisposeObject(StackValue svPtr, Executive exe)
+        /// color of this object to black.
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
```
What is this returning?

lukechurch commented on src/Engine/ProtoCore/DSASM/Heap.cs:
```diff
@@ -520,14 +570,262 @@ private void GCDisposeObject(StackValue svPtr, Executive exe)
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        private int PropagateMark(StackValue value)
```
Is this PropagateMark or RecursiveMark?

If Propagate, what is being propagated?

lukechurch commented on src/Engine/ProtoCore/DSASM/Heap.cs:
```diff
@@ -520,14 +570,262 @@ private void GCDisposeObject(StackValue svPtr, Executive exe)
+            }
+        }
+
+        public bool IsWaitingForRoots()
```
Property?

lukechurch commented on test/Engine/ProtoTest/DSASM/HeapCorruptionTests.cs:
```diff
@@ -1,86 +0,0 @@
-            Assert.IsFalse(thisTest.GetTestRuntimeCore().Heap.IsHeapCyclic());
-        }
-
-        [Test]
```
Why are these tests removed?

ke-yu replied:
These are for reference counting where heap could be circular referenced.

lukechurch commented:
Great work @ke-yu 

I've made a few comments, also

- I'd like to see a more agressive test suite being run that run a forced GC after every instruction and verifying no behaviour change

- We need a task for adding the memory benchmarks to every run

- Please verify that this doesn't cause regressions with the Revit test suite

- What do execution times look like for all of these suites?

What are you thinking about when to remove the cond-flag?

Let me know if you'd like another review when you've addressed the comments.

ke-yu commented:
@lukechurch , I've addressed your comments and enabled tri-color GC by default. 

Regarding to execution time, I did for tests: ProtoTest and DynamoCoreTests, the result shows in the pictures that I posted. They both run slighter faster than current GC implementation (235 seconds v.s. 247 seconds for ProtoTest; 676 seconds v.s. 691 seconds for DynamoCoreTests).

For other test libraries, they have less test cases, so I didn't do performance benchmark for them. 

For your first point, "run a forced GC after each instruction", the new implementation runs GC after *each instruction*, but depending on the memory allocation speed, the heap may skip GC. Do you mean running a full GC cycle after each instruction?

ke-yu commented:
@lukechurch ,memory profiling for DynamoCoreTests added.

lukechurch commented:
>For your first point, "run a forced GC after each instruction", the new implementation runs GC after each instruction, but depending on the memory allocation speed, the heap may skip GC. Do you mean running a full GC cycle after each instruction?

Yes. I meant running a forced GC after each instruction.

But overall this lgtm. Thanks. Lets bring it in and carefully watch for issues.
